File: dwcj-engine/src/main/java/org/dwcj/component/html/elements/Span.java
package org.dwcj.component.html.elements;

import org.dwcj.component.Component;
import org.dwcj.component.element.annotation.NodeName;
import org.dwcj.component.html.HtmlComponentContainer;

/**
 * Component representing a {@code span} element.
 *
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span">HTML span Tag</a>
 *
 * @author Hyyan Abo Fakher
 * @since 23.06
 */
// We're purposefully ignoring the deep inheritance warning here because we've designed our class
// hierarchy to meet the unique requirements of our UI framework. This design closely aligns with
// our framework's specific goals and emphasizes the need for caution when considering any changes.
//
// Any changes to the inheritance structure should be thoughtfully evaluated in the context of our
// framework's needs. The current structure is essential for meeting those needs.
@SuppressWarnings("squid:S110")
@NodeName("span")
public class Span extends HtmlComponentContainer<Span> {

  /**
   * Creates a new empty span.
   */
  public Span() {
    super();
  }

  /**
   * Creates a new span with the given text.
   *
   * @param text the text
   */
  public Span(String text) {
    super();
    setText(text);
  }

  /**
   * Creates a new span with the given child components.
   *
   * @param components the child components
   */
  public Span(Component... components) {
    super(components);
  }
}

File: webforj-kotlin/src/main/kotlin/com/webforj/kotlin/dsl/component/html/elements/Span.kt
package com.webforj.kotlin.dsl.component.html.elements

import com.webforj.component.html.elements.Span
import com.webforj.concern.HasComponents
import com.webforj.kotlin.dsl.WebforjDsl
import com.webforj.kotlin.dsl.init

/**
 * Creates an `Span` with an optional [text] value.
 * ```
 * ... {
 *  Span() // Empty Span element
 *  Span("text") // Span element with text
 * }
 * ```
 * @param text Optional text to add to the `Span`.
 * @param block The initialization steps of the `Span`.
 * @return The configured `Span` instance.
 * @see Span
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span">HTML span Tag</a>
 */
fun @WebforjDsl HasComponents.span(text: String? = null, block: @WebforjDsl Span.() -> Unit = {}): Span {
    val span = text?.let { Span(text) } ?: Span()
    return init(span, block)
}
